rfc/bulksave.md

```diff
@@ -0,0 +1,220 @@
+ This reduces the performance of front50 save api.
+
+ We improve the front50 save performance by using bulk insert query.
+ This will help save many pipelines at once which greatly improves the performance of front50.
+ We need to write a bulk save api which accepts an list of pipelines in a single rest api call
+ and save them in front50 using bulk insert query.
+
+### Goals and Non-Goals
+
+ We improve the front50 save performance by using bulk insert query.
+ This will help save many pipelines at once which greatly improves the performance of front50.
+ We need to write a bulk save api which accepts an list of pipelines in a single rest api call
+ and save them in front50 using bulk insert query.
+
+## Motivation and Rationale
+
+To improve the performance of multiple saving of pipelines at once using bulk insert sql query.
+
+## Timeline
+
+Already Implemented
+
+## Design
+
+1. Deck calls the gate api:
+
+   1.1	url: 		http://<ip address>:8084/pipelines/bulksave
+   method: 	POST
+   payload:	[{"keepWaitingPipelines":false,"limitConcurrent":true,"application":"apptest","spelEvaluator":"v4",
+   "lastModifiedBy":"admin","name":"pipeline_test","stages":[{"refId":"1","requisiteStageRefIds":[],
+   "type":"wait","name":"Wait","waitTime":30}],"index":3,"id":"66a5c77b-3998-495d-ad61-117313c5800c",
+   "triggers":[],"updateTs":"1599652266000"},{"keepWaitingPipelines":false,"limitConcurrent":true,
+   "application":"apptest","spelEvaluator":"v4","lastModifiedBy":"admin","name":"pipeline_test1",
+   "stages":[{"refId":"1","requisiteStageRefIds":[],"type":"wait","name":"Wait","waitTime":30}],
+   "index":4,"id":"66a5c77b-3998-495d-ad61-117313c5800c","triggers":[],"updateTs":"1599652266000"}......]
+
+
+2. Gate api flow:
+
+   2.1	The post request sent by the deck will hit the PipelineController Rest Controller endpoint.
+
+        com.netflix.spinnaker.gate.controllers.PipelineController.bulksavePipeline(Body pipeline..)
+   2.2 	TaskService will create a new task(synchronous job) for saving the pipeline.
+        taskService.createAndWaitForCompletion(operation)
+
+   2.3	The above new task will call the orca service post api to save the pipeline.
+
+        orcaService.doOperation(@Body Map<String, ? extends Object> body)
+        url: 		http://<ip address>:8083/ops
+        method: 	POST
+
+3.  Orca api flow:
+
+    3.1	The post request sent by the gate will hit the OperationsController Rest Controller endpoint.
+
+        com.netflix.spinnaker.orca.controllers.OperationsController Map<String, String> ops(@RequestBody Map input) {
+
+    3.2	The above method will start a new task with QueueExecutionRunner.kt.
+    3.3	The QueueExecutionRunner will call the StartExecutionHandler.kt.
+    3.4	The StartExecutionHandler.kt will call the SavePipelineTask.java.
+    3.5	SavePipelineTask will call the front 50 rest api.
+        Response response = front50Service.savePipelineList(List<pipeline>);
```

Comment:
`savePipelineList` would be better as `savePipelines`

Reply:
done